### PR TITLE
Add support for Tilt Zprobe

### PR DESCRIPTION
**Pull request by stratop80**
Hello,

I designed this type of z-probe, wrote the needed code, and tested successfully for several weeks up to now. It can be seen in http://www.thingiverse.com/thing:1427575, mounted in a self-designed Ultimaker derivative but with usual Arduino/RAMPS electronics.

This kind of z-probing is based on a twisting probe, that, on the first `G28` when reaching a specially defined `Z_SAFE_HOMING` point, deploys as it moves towards the bed, and performs an initial z-homing- After that, it moves to the center of the bed, where it homes for real, and then goes back to the safe homing point to stow the probe. Also, in `Marlin_main.cpp`, `deploy_probe` and `stow_probe` routines are altered.

I think it may inspire  similar pieces where the first deployment is made automatically and no servos or other parts are need. In the `Configuration.h` and `Marlin_main.cpp` files from the `RCBugFix` branch of my repository I have documented the proposed changes to give support to such devices.

I'm not sure if performing this pull request is the right way to do it, and so, I apologize if that`s the case. Feel free to take any needed action with this proposal. And thanks for all this work.

Regards.

### CODE REVIEW

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -2655,6 +2729,34 @@ inline void gcode_G28() {
...
   #endif
 
+  #if ENABLED(Z_PROBE_TILT)
+  
+// Performs HOMEAXIS(Z) in the middle of the bed
+
+      feedrate = XY_TRAVEL_SPEED;
+      destination[X_AXIS] = 100;
+      destination[Y_AXIS] = 100;
+      destination[Z_AXIS] = 20;
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Is that the middle of YOUR bed?

**Comment by stratop80** [comment]
Yes, my bed is 200x200, and the Z=20 is intended to start the HOMEAXIS(Z) with the probe deployed with no risk of crash.
But of course,  I suppose that the three coordinates should be user defined variables, if thats what you mean.
Let me know in any case, please. Thanks!

**Comment by stratop80** [comment]
I guess something like this would be much more sensible:

``` cpp
destination[X_AXIS] = (X_MIN_POS + X_MAX_POS) / 2;
destination[Y_AXIS] = (Y_MIN_POS + Y_MAX_POS) / 2;
destination[Z_AXIS] = Z_RAISE_BEFORE_PROBING;
```

just tell me.

**Comment by jbrazio** [comment]
Now that inductive sensors are mainstream using something as mechanically complicated as this and requiring Marlin to handle it in a special way.. well it's a cool exercise but I'm a bit unsure if it shall be incorporated into Marlin.

**Comment by stratop80** [comment]
Of course, I just leave there for that kind of consideration. But I must say that I've designed and used servo zprobes, that also used to be very popular too and not less complicated, and lately, inductive and capacitive sensors, that need dedicated additional boards or components to handle, only to find there was no way of having consistent measures. Servos suffer from electrical interferences, and inductive and capacitive sensors get affected by the hotbed temperature. Ive been strugling with it for months, and finally this thing simply works fine for me. In any case, thanks for commenting!

**Comment by AnHardt** [comment]
How do you find your device, to deploy the probe for homing, without already knowing z?

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -2655,6 +2729,34 @@ inline void gcode_G28() {
     }
   #endif
 
+  #if ENABLED(Z_PROBE_TILT)
+  
+// Performs HOMEAXIS(Z) in the middle of the bed
+
+      feedrate = XY_TRAVEL_SPEED;
+      destination[X_AXIS] = (X_MIN_POS + X_MAX_POS) / 2;
```
Comment (stratop80) [inline Marlin/Marlin_main.cpp]: corrected definition of the homing point

File: Marlin/Marlin_main.cpp [hunk 3]
```diff
@@ -2655,6 +2729,34 @@ inline void gcode_G28() {
...
+      destination[Y_AXIS] = (Y_MIN_POS + Y_MAX_POS) / 2;
+      destination[Z_AXIS] = Z_RAISE_BEFORE_PROBING;
```
Comment (stratop80) [inline Marlin/Marlin_main.cpp]: corrected definition of the homing point

**Comment by stratop80** [comment]
@AnHardt, It's because of redefining z_safe_homing point to where the special part that interacts with the probe lies;  the head moves now to that XY point with height enough (by means of `MIN_Z_HEIGHT_FOR_HOMING`) to be on the safe side. Then the normal `G28` z-homing begins, and in that process, while head is descending, the probe automatically deploys, taking a first z value. As this point (the special part) is defined out of the print area, z-homing is made on that part, whose height is known and different from that of the bed. But as we have yet a first z-value —differing on a fixed quantity (tenths of mm, in my setup, but can be whatever)— head can already be ordered to reach any point; So, then, `G28`  is ordered, —whenever the variable is set, of course— to perform a definitive z-homing in the middle of the bed, stow the probe and finish moving the head to 0,0,0. Further calls to `deploy_probe` or `stow_probe` work with proper z data, so from then on, routines like `G29` work seamlessly, as far as I have tested.
In many senses it may be similar to other systems, but it is fully automatic, and doesn't involve any special (electronic) part.
Hope I may have explained myself, and thanks for having a look in this!

**Comment by thinkyhead** [comment]
Typically, when someone — say, a Thingiverse user — makes a new peripheral that needs to work with Marlin (and other firmware), they will fork Marlin (and the other firmwares) on their Github account, add their custom support code to these forks, and then they will link to these forks from their Thingiverse page.

They will make a set of branches – one based on each release version of Marlin. For example `my_feature_for_marlin_1.0.2-1`, `my_feature_for_marlin_1.0.1`, `my_feature_for_marlin_1.1.0`. Then they might keep up to date with development, making branches such as `my_feature_for_marlin_dev` and `my_feature_for_marlin_rc3`.

If they're keeping up with development, they should update their branches by periodically rebasing against the latest versions of Marlin. For the Release Versions of Marlin, they can distribute their feature as a `MyFeature.patch` file. This way they never need to have the feature formally added to the Marlin project, and yet they can make sure everyone with their device is supported.

Since there are differing opinions on who should be supporting features for Marlin, some of the geeks around here will need to review this more to see how it fits our master plan.

Anyway, I would ask that you make this PR against the `MarlinDev/dev` branch also, because the `RC` branches are going to be short-lived, while the `dev` branch will be seeing a lot of action in the next phase. So we will need to move discussion over there soon.

File: Marlin/Marlin_main.cpp (hunk 3 above)
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: That's much better, thanks!

**Comment by stratop80** [comment]
Thanks for explaining all that, thinkyhead. I wasnt at all sure about what to do with this so I stayed quiet, as my handling of github (and many other things) is really primitive. I'll try to follow your directions. Be sure that my only interest is to share this, so its completely up to you devs how to use it, integrate or modify or not. I'll try to set up my branch(es), so I suppose you may close this PR now, or tell me if I should do anything apart from what its told. And again, thank you all devs for your comments and the effort you take here, for the benefit of all. Regards.

**Pull request closed without merging**